### PR TITLE
Implemented bp:innodb-page-fragmentation-counters-5.7 "InnoDB page fragmentation counters (5.7)"

### PR DESCRIPTION
**Pull request by percona-ysorokin**
https://blueprints.launchpad.net/percona-server/+spec/innodb-page-fragmentation-counters-5.7

Added the following InnoDB status variables:
'innodb_scan_pages_contiguous' - number of contiguous page reads.
'innodb_scan_pages_jumpy' - number of jumpy page reads.
'innodb_scan_data_in_pages' - amount of data in all pages read by
the latest query.
'innodb_scan_garbages_in_pages' - amount of garbage in all pages read
by the latest query.
'innodb_scan_pages_total_jump_distance' - total jump distance performed by
the latest query.

### CODE REVIEW

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
See 5.6

**Comment by percona-ysorokin** [comment]
Updated Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1254/

File: storage/innobase/srv/srv0srv.cc [hunk 1]
```diff
@@ -3390,3 +3392,48 @@ srv_is_undo_tablespace(
...
+srv_get_fragmentation_stats(fragmentation_stats_t& stats)
+{
+	THD*	thd = current_thd;
+	if (UNIV_LIKELY(thd != NULL)) {
+		ulonglong	local_scan_pages_contiguous = 0;
+		ulonglong	local_scan_pages_disjointed = 0;
+		ulonglong	local_scan_pages_total_seek_distance = 0;
+		ulonglong	local_scan_data_in_pages = 0;
+		ulonglong	local_scan_garbage_in_pages = 0;
+		thd_get_fragmentation_stats(thd,
```
Comment (laurynas-biveinis) [inline storage/innobase/srv/srv0srv.cc]: This usage with sql/sql_thd_internal_api.cc change makes me think that the struct is better moved to server layer?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/innobase/srv/srv0srv.cc (hunk 1 above)
Comment (percona-ysorokin) [reply storage/innobase/srv/srv0srv.cc]: Reworked

**Comment by percona-ysorokin** [comment]
Another Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1273/

**Review by laurynas-biveinis — COMMENTED** [review]
Same comment as for 5.6

**Comment by percona-ysorokin** [comment]
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1277/

File: sql/sql_thd_internal_api.h [hunk 2]
```diff
@@ -99,4 +99,34 @@ bool is_mysql_datadir_path(const char *path);
 
 int mysql_tmpfile_path(const char *path, const char *prefix);
 
+/** Page gragmentation statistics */
```
Comment (laurynas-biveinis) [inline sql/sql_thd_internal_api.h]: fragmentation

**Review by laurynas-biveinis — APPROVED** [review]
Fix a comment typo and push

File: sql/sql_thd_internal_api.h (hunk 2 above)
Comment (percona-ysorokin) [reply sql/sql_thd_internal_api.h]: fixed

**Pull request merged**